service-base/src/main/java/org/eclipse/hono/service/impl/RequestResponseEndpoint.java

```diff
@@ -0,0 +1,147 @@
+     */
+    protected abstract boolean verifyMessage(final ResourceIdentifier linkTarget, final Message msg);
+
+    /**
+     * Process the received AMQP message.
+     *
+     * @param message The Message to process.
+     */
+    protected abstract void processRequest(final Message message);
+
+    /**
+     * Construct an AMQP reply message that is send back to the caller.
+     *
+     * @param message The Message as JsonObject from which a reply message is constructed.
+     */
+    protected abstract Message getAmqpReply(final io.vertx.core.eventbus.Message<JsonObject> message);
+
+    /**
+     * Gets the credits this endpoint grants.
+     *
+     * @return The amount of credits granted.
+     */
+    public final int getPrefetchReceiver() {
+        return prefetchReceiver;
+    }
+
+    /**
+     * Set the credits this endpoint shall grant.
+     */
+    protected final void setPrefetchReceiver(final int prefetchReceiver) {
+        this.prefetchReceiver = prefetchReceiver;
+    }
+
+    @Override
+    public void onLinkAttach(final ProtonReceiver receiver, final ResourceIdentifier targetAddress) {
+        if (ProtonQoS.AT_MOST_ONCE.equals(receiver.getRemoteQoS())) {
+            logger.debug("client wants to use AT MOST ONCE delivery mode for {} endpoint, this is not supported.",getName());
+            receiver.setCondition(condition(AmqpError.PRECONDITION_FAILED.toString(), "endpoint requires AT_LEAST_ONCE QoS"));
+            receiver.close();
+        } else {
+
+            logger.debug("establishing link for receiving messages from client [{}]", MessageHelper.getLinkName(receiver));
+            receiver
+                    .setQoS(ProtonQoS.AT_LEAST_ONCE)
+                    .setAutoAccept(true) // settle received messages if the handler succeeds
+                    .setPrefetch(prefetchReceiver)
+                    .handler((delivery, message) -> {
+                        if (verifyMessage(targetAddress, message)) {
+                            try {
+                                processRequest(message);
+                            } catch (DecodeException e) {
+                                MessageHelper.rejected(delivery, AmqpError.DECODE_ERROR.toString(), "malformed payload");
+                            }
+                        } else {
+                            MessageHelper.rejected(delivery, AmqpError.DECODE_ERROR.toString(), "malformed message");
+                            // we close the link if the client sends a message that does not comply with the API spec
+                            onLinkDetach(receiver, condition(AmqpError.DECODE_ERROR.toString(), "invalid message received"));
+                        }
+                    }).closeHandler(clientDetached -> onLinkDetach(clientDetached.result()))
+                    .open();
+        }
+    }
+
+    @Override
+    public void onLinkAttach(final ProtonSender sender, final ResourceIdentifier targetResource) {
```

Comment:
is this method supposed to be overridden? if so, please document what this method currently does and suggest to subclasses to override it. Otherwise, this method should be final.